<a href="https://colab.research.google.com/github/nmquintero/MLandDS/blob/main/Data%20Preprocessing/Data_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Sprint 10: Data Preprocessing

1. **Data Cleaning Techniques and Handling Missing Values**
   - **Introduction to Data Cleaning:**
     - The importance of data cleaning in the data preprocessing pipeline.
     - Common data quality issues: duplicates, incorrect data types, outliers, and inconsistencies.
   - **Handling Missing Values:**
     - Understanding types of missing data: MCAR (Missing Completely at Random), MAR (Missing at Random), MNAR (Missing Not at Random).
     - Techniques for handling missing values:
       - **Removing Missing Data:** When and how to drop rows or columns with missing values.
       - **Imputation Methods:**
         - Simple imputation: Mean, median, and mode.
         - Advanced imputation: KNN imputation, multivariate imputation, and regression imputation.
       - **Creating Indicator Variables:** Use of binary flags to indicate the presence of missing data.

2. **Data Scaling and Normalization**
   - **Importance of Data Scaling:**
     - How scaling affects the performance of machine learning algorithms, especially those based on distance or gradient descent.
   - **Scaling Techniques:**
     - **Min-Max Scaling:** Rescaling data to a fixed range, typically 0 to 1.
     - **Z-Score Normalization (Standardization):** Adjusting data to have a mean of 0 and a standard deviation of 1.
     - **Robust Scaling:** Scaling using the median and interquartile range, effective for data with outliers.
   - **When to Use Each Technique:** Practical guidelines for choosing the appropriate scaling method.

3. **Introduction to Feature Engineering**
   - **Definition and Importance:**
     - Transforming raw data into features that better represent the underlying problem to the predictive models.
   - **Techniques in Feature Engineering:**
     - **Creating New Features:** Deriving new features from existing ones (e.g., combining or transforming variables).
     - **Encoding Categorical Variables:** Techniques like one-hot encoding, label encoding, and target encoding.
     - **Interaction Features:** Creating features by combining multiple variables to capture their interaction.
     - **Polynomial Features:** Generating polynomial combinations of features to model non-linear relationships.



### Few words about scikit-learn

We will use **`scikit-learn`**, often abbreviated as `sklearn`, - one of the most popular and powerful open-source libraries in Python for machine learning and data analysis. It provides a comprehensive suite of tools for building, evaluating, and deploying machine learning models, making it an essential tool for data scientists, machine learning engineers, and statisticians.

##### **Key Features of `scikit-learn`:**

1. **Machine Learning Algorithms:**
   - `scikit-learn` includes a wide array of machine learning algorithms for both supervised and unsupervised learning tasks. This includes algorithms for:
     - **Classification:** Logistic Regression, Support Vector Machines (SVM), Random Forests, k-Nearest Neighbors (k-NN), Naive Bayes, etc.
     - **Regression:** Linear Regression, Ridge Regression, Lasso, Decision Trees, etc.
     - **Clustering:** K-Means, Hierarchical Clustering, DBSCAN, etc.
     - **Dimensionality Reduction:** Principal Component Analysis (PCA), t-SNE, etc.
     - **Model Selection and Evaluation:** Tools like Grid Search, Cross-Validation, and metrics such as accuracy, precision, recall, F1-score, and more.

2. **Preprocessing Tools:**
   - `scikit-learn` offers a variety of preprocessing utilities for preparing data before it is fed into a machine learning model. This includes:
     - **Data Scaling and Normalization:** `StandardScaler`, `MinMaxScaler`, `RobustScaler`.
     - **Encoding Categorical Variables:** `OneHotEncoder`, `LabelEncoder`.
     - **Feature Engineering:** Polynomial features, interaction terms, and more.
     - **Imputation:** `SimpleImputer`, `KNNImputer` for handling missing data.
   
3. **Model Validation and Selection:**
   - `scikit-learn` provides powerful tools for model validation and hyperparameter tuning, ensuring that your models generalize well to unseen data.
     - **Cross-Validation:** Techniques like k-fold cross-validation, stratified k-fold, and more.
     - **Grid Search and Random Search:** `GridSearchCV`, `RandomizedSearchCV` to find the best hyperparameters for your models.
   
4. **Pipelines:**
   - Pipelines in `scikit-learn` allow you to streamline the process of creating machine learning workflows by combining multiple steps (e.g., preprocessing, model fitting) into a single, cohesive process.
     - **Example:**
       ```python
       from sklearn.pipeline import Pipeline
       from sklearn.preprocessing import StandardScaler
       from sklearn.linear_model import LogisticRegression

       pipeline = Pipeline([
           ('scaler', StandardScaler()),
           ('classifier', LogisticRegression())
       ])

       pipeline.fit(X_train, y_train)
       ```

5. **Model Evaluation Metrics:**
   - `scikit-learn` includes a rich set of metrics for evaluating model performance, from standard accuracy to more complex metrics like ROC AUC, mean squared error (MSE), and adjusted R².
     - **Example:**
       ```python
       from sklearn.metrics import accuracy_score, confusion_matrix

       y_pred = pipeline.predict(X_test)
       accuracy = accuracy_score(y_test, y_pred)
       cm = confusion_matrix(y_test, y_pred)
       ```

6. **Ease of Use:**
   - `scikit-learn` is designed with simplicity in mind, using consistent and intuitive APIs. It integrates seamlessly with other popular Python libraries such as NumPy, SciPy, and Pandas, allowing for easy data manipulation and visualization.
   
7. **Extensive Documentation and Community Support:**
   - The library is well-documented, with extensive tutorials, examples, and user guides available. It also has a large and active community, making it easy to find help and resources.

##### **Use Cases:**

- **Data Preprocessing:** Preparing raw data for analysis, including handling missing values, scaling, and encoding.
- **Building Predictive Models:** Developing models for tasks such as predicting customer churn, classifying images, or forecasting sales.
- **Feature Engineering:** Enhancing the predictive power of datasets by creating new features or transforming existing ones.
- **Model Evaluation:** Assessing the performance of machine learning models to ensure they generalize well to new data.
- **Hyperparameter Tuning:** Finding the best parameters for machine learning models to optimize their performance.



##1. Data Cleaning Techniques and Handling Missing Values

In the real world, raw data is rarely perfect. It often contains errors, inconsistencies, and missing values. Data cleaning is the process of detecting and correcting these issues to improve the quality of the data, which is crucial for building reliable machine learning models.

###1.1. Common Data Issues
- **Duplicates:** These are repeated entries in the dataset. Duplicate rows can skew your analysis or model, especially if they represent the same event or entity.
- **Incorrect Data Types:** Data that is incorrectly typed (e.g., numbers stored as strings) can lead to errors in analysis and model training.
- **Outliers:** Extreme values that are far removed from the rest of the data. They can distort the results of statistical analyses or machine learning models.

###1.2. Handling Missing Values
Missing data can occur for various reasons, such as data entry errors, equipment malfunctions, or privacy concerns. It's essential to handle missing values carefully, as improper handling can lead to biased results.

**Types of Missing Data:**
- **MCAR (Missing Completely at Random):** The probability of data being missing is independent of the data itself.
- **MAR (Missing at Random):** The probability of missing data is related to other observed data, but not to the missing data itself.
- **MNAR (Missing Not at Random):** The missingness is related to the unobserved data itself.



###1.3. Techniques for Handling Missing Values


- **Removing Missing Data:** Dropping rows or columns with missing values is the simplest approach, but it can lead to loss of valuable information.

In [50]:
#You need this code block only in case using Google Colaboratory
from google.colab import drive
drive.mount('/content/drive')
%cd /content/drive/Othercomputers/Mi MacBook Air (1)/MASTER IA/Data Preprocessing

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
/content/drive/Othercomputers/Mi MacBook Air (1)/MASTER IA/Data Preprocessing


In [75]:
import pandas as pd

df = pd.read_csv('data.csv')
df_cleaned = df.dropna()  # Drops all rows with any missing values
# Drop rows with any missing values in numerical columns
df = df.dropna(subset=['feature1', 'feature2'])

In [76]:
df.head(4)

,feature1,feature2,feature,height,weight,date,color,target,age,income,x,categorical_column
0,25.5,30.2,10,1.60,55,2024-01-01,red,1,25,30000,2,low
3,45.0,55.3,150,1.90,90,2024-03-10,red,1,55,90000,5,very_high
4,18.7,22.4,200,1.68,60,2024-07-22,blue,1,28,35000,6,low
5,60.2,45.1,10,1.72,68,2024-02-14,green,0,32,48000,7,medium


In [77]:
df_cleaned.head(4)

,feature1,feature2,feature,height,weight,date,color,target,age,income,x,categorical_column
0,25.5,30.2,10,1.60,55,2024-01-01,red,1,25,30000,2,low
3,45.0,55.3,150,1.90,90,2024-03-10,red,1,55,90000,5,very_high
4,18.7,22.4,200,1.68,60,2024-07-22,blue,1,28,35000,6,low
5,60.2,45.1,10,1.72,68,2024-02-14,green,0,32,48000,7,medium


**Imputation Methods:**
  - **Mean Imputation:** Replace missing values with the mean of the column.


In [78]:
# Impute missing values with the mean of the respective column
df['feature1'].fillna(df['feature1'].mean(), inplace=True)
df['feature2'].fillna(df['feature2'].mean(), inplace=True)

/tmp/ipykernel_9249/2010758680.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['feature1'].fillna(df['feature1'].mean(), inplace=True)
/tmp/ipykernel_9249/2010758680.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=Tr

In [79]:
df.head(4)

,feature1,feature2,feature,height,weight,date,color,target,age,income,x,categorical_column
0,25.5,30.2,10,1.60,55,2024-01-01,red,1,25,30000,2,low
3,45.0,55.3,150,1.90,90,2024-03-10,red,1,55,90000,5,very_high
4,18.7,22.4,200,1.68,60,2024-07-22,blue,1,28,35000,6,low
5,60.2,45.1,10,1.72,68,2024-02-14,green,0,32,48000,7,medium


  - **Median Imputation:** Useful when the data is skewed.




In [80]:
df['feature1'].fillna(df['feature1'].median(), inplace=True)

/tmp/ipykernel_9249/845216444.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['feature1'].fillna(df['feature1'].median(), inplace=True)


In [81]:
df.head(4)

,feature1,feature2,feature,height,weight,date,color,target,age,income,x,categorical_column
0,25.5,30.2,10,1.60,55,2024-01-01,red,1,25,30000,2,low
3,45.0,55.3,150,1.90,90,2024-03-10,red,1,55,90000,5,very_high
4,18.7,22.4,200,1.68,60,2024-07-22,blue,1,28,35000,6,low
5,60.2,45.1,10,1.72,68,2024-02-14,green,0,32,48000,7,medium


  - **Mode Imputation:** Often used for categorical data.

In [82]:
df['feature1'].fillna(df['feature1'].mode()[0], inplace=True)

/tmp/ipykernel_9249/1991520260.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['feature1'].fillna(df['feature1'].mode()[0], inplace=True)


In [83]:
df.head(4)

,feature1,feature2,feature,height,weight,date,color,target,age,income,x,categorical_column
0,25.5,30.2,10,1.60,55,2024-01-01,red,1,25,30000,2,low
3,45.0,55.3,150,1.90,90,2024-03-10,red,1,55,90000,5,very_high
4,18.7,22.4,200,1.68,60,2024-07-22,blue,1,28,35000,6,low
5,60.2,45.1,10,1.72,68,2024-02-14,green,0,32,48000,7,medium


- **Advanced Imputation:**
  - **K-Nearest Neighbors (KNN) Imputation:** Estimates missing values based on the values of the nearest neighbors.
  - **Multivariate Imputation:** Considers the relationships between features to impute missing values.

**Creating Indicator Variables:**
Another approach is to create a binary indicator variable that flags the presence of missing data.


In [84]:
df['missing_column'] = df['feature1'].isnull().astype(int)

In [85]:
df.head(4)

,feature1,feature2,feature,height,weight,date,color,target,age,income,x,categorical_column,missing_column
0,25.5,30.2,10,1.60,55,2024-01-01,red,1,25,30000,2,low,0
3,45.0,55.3,150,1.90,90,2024-03-10,red,1,55,90000,5,very_high,0
4,18.7,22.4,200,1.68,60,2024-07-22,blue,1,28,35000,6,low,0
5,60.2,45.1,10,1.72,68,2024-02-14,green,0,32,48000,7,medium,0


## 2. Data Scaling and Normalization

In machine learning, data scaling and normalization are crucial preprocessing steps that can significantly affect the performance of your models. These techniques are particularly important for algorithms that rely on distance metrics (e.g., K-Nearest Neighbors, Support Vector Machines) or gradient-based optimization methods (e.g., linear regression, neural networks). Below, we will explore why scaling is important, different scaling techniques, and when to use each technique.

###2.1. Importance of Data Scaling

Data scaling ensures that all features contribute equally to the model's learning process. When features have vastly different ranges, those with larger ranges can dominate the distance calculations or the gradient steps, leading to suboptimal models. For instance:

- **Distance-based Algorithms:** In algorithms like K-Nearest Neighbors (KNN) or Support Vector Machines (SVM), the distance between data points is crucial. If one feature has a much larger range than others, it will disproportionately influence the distance metric, potentially skewing the results.
  
- **Gradient Descent:** Algorithms like linear regression and neural networks use gradient descent to minimize a cost function. If the features are not scaled, the gradient descent algorithm may take longer to converge or may converge to a suboptimal solution due to the uneven steps taken in the parameter space.

###2.2. Scaling Techniques

#### **Min-Max Scaling**

**Min-Max Scaling** (or normalization) rescales the data to a fixed range, usually [0, 1]. This technique is particularly useful when you know that your data follows a distribution with clear upper and lower bounds.

**Formula:**
$$
X' = \frac{X - X_{\text{min}}}{X_{\text{max}} - X_{\text{min}}}
$$
Where:
- $X$ is the original value.
- $X_{\text{min}}$ and $X_{\text{max}}$ are the minimum and maximum values of the feature.

**Example:**

Suppose you have a feature with values ranging from 10 to 200. Applying Min-Max Scaling to this feature will rescale it to a range of [0, 1].

In [86]:
from sklearn.preprocessing import MinMaxScaler
import pandas as pd

df = pd.read_csv('data.csv')

# Select only numerical columns for scaling
numerical_df = df.select_dtypes(include=['number'])

scaler = MinMaxScaler()
scaled_data = scaler.fit_transform(numerical_df)

print(scaled_data)

[[0.18735363 0.20487805 0.         0.         0.07894737 1.
  0.09090909 0.07692308 0.        ]
 [0.34192037        nan 0.21052632 0.5        0.73684211 0.
  0.39393939 0.38461538 0.125     ]
 [       nan        nan 0.47368421 0.73333333 0.52631579 0.
  0.6969697  0.69230769 0.25      ]
 [0.6440281  0.81707317 0.73684211 1.         1.         1.
  1.         1.         0.375     ]
 [0.02810304 0.01463415 1.         0.26666667 0.21052632 1.
  0.18181818 0.15384615 0.5       ]
 [1.         0.56829268 0.         0.4        0.42105263 0.
  0.3030303  0.35384615 0.625     ]
 [0.38173302        nan 0.21052632 0.83333333 0.86842105 0.
  0.54545455 0.56923077 0.75      ]
 [0.55737705 0.41219512 0.47368421 0.6        0.57894737 1.
  0.48484848 0.46153846 0.875     ]
 [0.28337237 0.28292683 0.73684211 0.16666667 0.15789474 0.
  0.24242424 0.24615385 1.        ]
 [0.87822014 0.93414634 1.         0.33333333 0.28947368 1.
  0.60606061 0.66153846 0.        ]
 [0.1147541  0.17317073 0.         0.933


This rescaled feature now lies between 0 and 1.

One more example of **Min-Max Scaling:**

In [87]:
# Select numerical columns for scaling
numerical_cols = df[['feature1', 'feature2']]

# Keep categorical and target columns separate
categorical_cols = df[['categorical_column', 'target']]


In [88]:
from sklearn.preprocessing import MinMaxScaler

# Initialize the scaler
scaler = MinMaxScaler()

# Fit and transform the numerical columns
numerical_scaled = scaler.fit_transform(numerical_cols)

# Convert the scaled data back to a DataFrame
numerical_scaled_df = pd.DataFrame(numerical_scaled, columns=numerical_cols.columns)


In [89]:
# Combine scaled numerical data with the categorical columns
df_scaled = pd.concat([numerical_scaled_df, categorical_cols.reset_index(drop=True)], axis=1)

# Display the final DataFrame
print(df_scaled)


    feature1  feature2 categorical_column  target
0   0.187354  0.204878                low       1
1   0.341920       NaN             medium       0
2        NaN       NaN               high       0
3   0.644028  0.817073          very_high       1
4   0.028103  0.014634                low       1
5   1.000000  0.568293             medium       0
6   0.381733       NaN               high       0
7   0.557377  0.412195          very_high       1
8   0.283372  0.282927                low       0
9   0.878220  0.934146             medium       1
10  0.114754  0.173171               high       0
11  0.482436  0.507317          very_high       1
12       NaN       NaN                low       0
13  0.709602  0.743902             medium       1
14  0.070258  0.070732               high       0
15  0.437939  0.456098          very_high       1
16  0.824356  0.660976                low       0
17  0.622951       NaN             medium       1
18  0.229508  0.268293               high       0


**Use Case:**
- Use Min-Max Scaling when you want all your features to have the same scale (e.g., in algorithms like KNN, SVM).
- Particularly useful when the data is distributed across a known and fixed range.


#### **Z-Score Normalization (Standardization)**

**Z-Score Normalization** (or standardization) rescales the data to have a mean of 0 and a standard deviation of 1. This technique is useful when the data follows a Gaussian (normal) distribution but not necessarily bounded.

**Formula:**
$$
X' = \frac{X - \mu}{\sigma}
$$
Where:
- $\mu$ is the mean of the feature.
- $\sigma$ is the standard deviation of the feature.

**Example:**

Suppose you have a feature with values [10, 50, 100, 150, 200]. Applying Z-Score Normalization will adjust the data to have a mean of 0 and a standard deviation of 1.

In [91]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
standardized_data = scaler.fit_transform(numerical_df)

print(standardized_data)

[[-1.02032742 -0.96688656 -1.2902465  -1.79679815 -1.41296754  1.
  -1.43265691 -1.52419956 -1.49378879]
 [-0.45321242         nan -0.70979331 -0.02059368  0.86601237 -1.
  -0.24093507 -0.31743227 -1.11204277]
 [        nan         nan  0.01577318  0.80830175  0.13673879 -1.
   0.95078677  0.88933502 -0.73029674]
 [ 0.65523964  1.2111572   0.74133968  1.7556108   1.77760433  1.
   2.14250862  2.09610232 -0.34855072]
 [-1.60462773 -1.64372884  1.46690617 -0.8494891  -0.95717156  1.
  -1.07514036 -1.22250773  0.03319531]
 [ 1.96132269  0.32605575 -1.2902465  -0.37583457 -0.22789799 -1.
  -0.59845162 -0.43810899  0.41494133]
 [-0.30713734         nan -0.70979331  1.16354264  1.32180835 -1.
   0.35492585  0.40662811  0.79668736]
 [ 0.33731153 -0.22930202  0.01577318  0.33464722  0.31905719  1.
   0.11658148 -0.01574044  1.17843338]
 [-0.66802871 -0.68920767  0.74133968 -1.20472999 -1.13948996 -1.
  -0.83679599 -0.86047755  1.56017941]
 [ 1.5145048   1.62767553  1.46690617 -0.61266183 -0.68


One more example of **Z-Score Normalization (Standardization):**

In [92]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
df_standardized = scaler.fit_transform(numerical_cols)

This rescaled feature now has a mean of 0 and a standard deviation of 1.

**Use Case:**
- Use Z-Score Normalization when your data follows a normal distribution.
- It's the default choice for many machine learning models, especially those assuming normally distributed data.



#### **Robust Scaling**
**Robust Scaling** uses the median and interquartile range (IQR) for scaling, which makes it robust to outliers. Unlike Min-Max Scaling and Z-Score Normalization, Robust Scaling is less influenced by extreme values in the data.

**Formula:**

$$X'=\frac{X - \text{median}(X)}{\text{IQR}(X)}$$

Where:
- $\text{IQR}(X)$ is the range between the 25th percentile ($Q1$) and the 75th percentile ($Q3$).

**Example:**

Suppose you have a feature with values [10, 50, 100, 150, 200]. Robust Scaling will adjust the data based on the median and $\text{IQR}$.

In [95]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()
robust_scaled_data = scaler.fit_transform(numerical_df)

print(robust_scaled_data)

[[-0.64850136 -0.5799458  -0.9        -1.05454545 -0.8         0.5
  -0.85185185 -0.97142857 -1.        ]
 [-0.28882834         nan -0.5         0.03636364  0.53333333 -0.5
  -0.11111111 -0.20952381 -0.75      ]
 [        nan         nan  0.          0.54545455  0.10666667 -0.5
   0.62962963  0.55238095 -0.5       ]
 [ 0.41416894  0.7804878   0.5         1.12727273  1.06666667  0.5
   1.37037037  1.31428571 -0.25      ]
 [-1.01907357 -1.00271003  1.         -0.47272727 -0.53333333  0.5
  -0.62962963 -0.78095238  0.        ]
 [ 1.24250681  0.22764228 -0.9        -0.18181818 -0.10666667 -0.5
  -0.33333333 -0.28571429  0.25      ]
 [-0.19618529         nan -0.5         0.76363636  0.8        -0.5
   0.25925926  0.24761905  0.5       ]
 [ 0.21253406 -0.11924119  0.          0.25454545  0.21333333  0.5
   0.11111111 -0.01904762  0.75      ]
 [-0.42506812 -0.40650407  0.5        -0.69090909 -0.64       -0.5
  -0.48148148 -0.55238095  1.        ]
 [ 0.95912807  1.04065041  1.         -0.32727

This scaling ensures that outliers have less influence on the scaled data.



One more example of **Robust Scaling:**

In [96]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()
df_robust_scaled = scaler.fit_transform(numerical_cols)


**Use Case:**
- Use Robust Scaling when your data contains outliers or follows a non-Gaussian distribution.
- Ideal for situations where the data is skewed or contains extreme values that should not heavily influence the scaling.

###2.3. When to Use Each Technique

- **Min-Max Scaling:**
  - Use when you want to bring all features into the same scale, especially for algorithms that rely on distance metrics (e.g., KNN, SVM).
  - Suitable for data with known bounds.

- **Z-Score Normalization (Standardization):**
  - Use when your data follows a normal distribution, or when using models that assume normally distributed data (e.g., linear regression, logistic regression).
  - It's also a good general-purpose scaler when you don't have specific needs for Min-Max Scaling or Robust Scaling.

- **Robust Scaling:**
  - Use when your data contains outliers or is heavily skewed.
  - Ideal for datasets with features that have varying scales and are not normally distributed.

Scaling and normalization are critical steps in the data preprocessing pipeline. The choice of technique depends on the nature of your data and the requirements of the machine learning algorithm you are using. By appropriately scaling your data, you ensure that your models perform optimally and converge more quickly during training.

**c. Choosing the Right Scaling Technique:**
- **Min-Max Scaling:** Use when you know the boundaries of your data or when the algorithm requires data in a specific range.
- **Z-Score Normalization:** Suitable for normally distributed data or when you want to standardize features to have equal importance.
- **Robust Scaling:** Best when dealing with data that has outliers.


##3. Introduction to Feature Engineering

Feature engineering is a crucial step in the data preprocessing pipeline. It involves transforming raw data into features that better represent the underlying problem to the predictive models. Good feature engineering can significantly improve the performance of machine learning models by providing them with the most relevant and informative input data.

###3.1. Definition and Importance

**Feature Engineering** refers to the process of selecting, modifying, or creating new features from raw data to improve the performance of machine learning models. The goal is to create features that capture the underlying patterns in the data more effectively than the original raw data. Proper feature engineering can lead to simpler models, improved accuracy, and reduced training times.

###3.2. Techniques in Feature Engineering

#### **Creating New Features**

Creating new features involves deriving new variables from existing ones that better represent the underlying structure of the data. This could involve mathematical transformations, aggregations, or domain-specific operations.

**Example 1: Mathematical Transformations**

Suppose you have a dataset with the `height` and `weight` of individuals. You can create a new feature, Body Mass Index (BMI), which is a better indicator of health than either `height` or `weight` alone.

In [98]:
df.head(4)

,feature1,feature2,feature,height,weight,date,color,target,age,income,x,categorical_column
0,25.5,30.2,10,1.60,55,2024-01-01,red,1,25,30000,2,low
1,32.1,NaN,50,1.75,80,2024-06-15,blue,0,35,50000,3,medium
2,NaN,NaN,100,1.82,72,2024-12-31,green,0,45,70000,4,high
3,45.0,55.3,150,1.90,90,2024-03-10,red,1,55,90000,5,very_high


In [99]:
# Creating a new feature: BMI
df['BMI'] = df['weight'] / (df['height'] ** 2)

print(df)

    feature1  feature2  feature  height  weight        date  color  target  \
0       25.5      30.2       10    1.60      55  2024-01-01    red       1   
1       32.1       NaN       50    1.75      80  2024-06-15   blue       0   
2        NaN       NaN      100    1.82      72  2024-12-31  green       0   
3       45.0      55.3      150    1.90      90  2024-03-10    red       1   
4       18.7      22.4      200    1.68      60  2024-07-22   blue       1   
5       60.2      45.1       10    1.72      68  2024-02-14  green       0   
6       33.8       NaN       50    1.85      85  2024-09-05    red       0   
7       41.3      38.7      100    1.78      74  2024-04-18   blue       1   
8       29.6      33.4      150    1.65      58  2024-11-30  green       0   
9       55.0      60.1      200    1.70      63  2024-08-12    red       1   
10      22.4      28.9       10    1.88      82  2024-05-25   blue       0   
11      38.1      42.6       50    1.80      78  2024-10-08  gre

**Example 2: Date-Time Features**

Suppose you have a `date` feature. You can create new features such as `day_of_week`, `month`, or `year` to capture patterns that occur over time.

In [100]:
# Convert to datetime
df['date'] = pd.to_datetime(df['date'])

# Creating new date-time features
df['day_of_week'] = df['date'].dt.day_name()
df['month'] = df['date'].dt.month
df['year'] = df['date'].dt.year

print(df)

    feature1  feature2  feature  height  weight       date  color  target  \
0       25.5      30.2       10    1.60      55 2024-01-01    red       1   
1       32.1       NaN       50    1.75      80 2024-06-15   blue       0   
2        NaN       NaN      100    1.82      72 2024-12-31  green       0   
3       45.0      55.3      150    1.90      90 2024-03-10    red       1   
4       18.7      22.4      200    1.68      60 2024-07-22   blue       1   
5       60.2      45.1       10    1.72      68 2024-02-14  green       0   
6       33.8       NaN       50    1.85      85 2024-09-05    red       0   
7       41.3      38.7      100    1.78      74 2024-04-18   blue       1   
8       29.6      33.4      150    1.65      58 2024-11-30  green       0   
9       55.0      60.1      200    1.70      63 2024-08-12    red       1   
10      22.4      28.9       10    1.88      82 2024-05-25   blue       0   
11      38.1      42.6       50    1.80      78 2024-10-08  green       1   

#### **Encoding Categorical Variables**

Machine learning algorithms generally require numerical input. Therefore, categorical variables must be converted into numerical format. There are several techniques to encode categorical variables:

**a. One-Hot Encoding**

One-hot encoding converts each category into a new binary column. This method is suitable when there are a limited number of categories.

**Example:**

In [101]:
# One-Hot Encoding
df_encoded = pd.get_dummies(df, columns=['color'])

print(df_encoded)

    feature1  feature2  feature  height  weight       date  target  age  \
0       25.5      30.2       10    1.60      55 2024-01-01       1   25   
1       32.1       NaN       50    1.75      80 2024-06-15       0   35   
2        NaN       NaN      100    1.82      72 2024-12-31       0   45   
3       45.0      55.3      150    1.90      90 2024-03-10       1   55   
4       18.7      22.4      200    1.68      60 2024-07-22       1   28   
5       60.2      45.1       10    1.72      68 2024-02-14       0   32   
6       33.8       NaN       50    1.85      85 2024-09-05       0   40   
7       41.3      38.7      100    1.78      74 2024-04-18       1   38   
8       29.6      33.4      150    1.65      58 2024-11-30       0   30   
9       55.0      60.1      200    1.70      63 2024-08-12       1   42   
10      22.4      28.9       10    1.88      82 2024-05-25       0   27   
11      38.1      42.6       50    1.80      78 2024-10-08       1   33   
12       NaN       NaN   

**b. Label Encoding**

Label encoding assigns a unique integer to each category. It is simple but assumes an ordinal relationship between the categories, which may not be appropriate for all cases.

**Example:**

In [102]:
from sklearn.preprocessing import LabelEncoder

# Label Encoding
encoder = LabelEncoder()
df['color_encoded'] = encoder.fit_transform(df['color'])

print(df)

    feature1  feature2  feature  height  weight       date  color  target  \
0       25.5      30.2       10    1.60      55 2024-01-01    red       1   
1       32.1       NaN       50    1.75      80 2024-06-15   blue       0   
2        NaN       NaN      100    1.82      72 2024-12-31  green       0   
3       45.0      55.3      150    1.90      90 2024-03-10    red       1   
4       18.7      22.4      200    1.68      60 2024-07-22   blue       1   
5       60.2      45.1       10    1.72      68 2024-02-14  green       0   
6       33.8       NaN       50    1.85      85 2024-09-05    red       0   
7       41.3      38.7      100    1.78      74 2024-04-18   blue       1   
8       29.6      33.4      150    1.65      58 2024-11-30  green       0   
9       55.0      60.1      200    1.70      63 2024-08-12    red       1   
10      22.4      28.9       10    1.88      82 2024-05-25   blue       0   
11      38.1      42.6       50    1.80      78 2024-10-08  green       1   

**c. Target Encoding**

Target encoding involves replacing each category with the mean of the target variable for that category. This technique is useful when the categorical variable has a large number of levels.

**Example:**

In [103]:

# Target Encoding
df['color_encoded'] = df.groupby('color')['target'].transform('mean')

print(df)

    feature1  feature2  feature  height  weight       date  color  target  \
0       25.5      30.2       10    1.60      55 2024-01-01    red       1   
1       32.1       NaN       50    1.75      80 2024-06-15   blue       0   
2        NaN       NaN      100    1.82      72 2024-12-31  green       0   
3       45.0      55.3      150    1.90      90 2024-03-10    red       1   
4       18.7      22.4      200    1.68      60 2024-07-22   blue       1   
5       60.2      45.1       10    1.72      68 2024-02-14  green       0   
6       33.8       NaN       50    1.85      85 2024-09-05    red       0   
7       41.3      38.7      100    1.78      74 2024-04-18   blue       1   
8       29.6      33.4      150    1.65      58 2024-11-30  green       0   
9       55.0      60.1      200    1.70      63 2024-08-12    red       1   
10      22.4      28.9       10    1.88      82 2024-05-25   blue       0   
11      38.1      42.6       50    1.80      78 2024-10-08  green       1   

#### **Interaction Features**

Interaction features are created by combining two or more variables to capture the interactions between them. These features can reveal relationships that are not apparent when considering the variables independently.

**Example:**

Suppose you have `age` and `income` as features. An interaction term could be created by multiplying these two features to capture how income changes with age.

In [104]:

# Creating an interaction feature
df['age_income_interaction'] = df['age'] * df['income']

print(df)

    feature1  feature2  feature  height  weight       date  color  target  \
0       25.5      30.2       10    1.60      55 2024-01-01    red       1   
1       32.1       NaN       50    1.75      80 2024-06-15   blue       0   
2        NaN       NaN      100    1.82      72 2024-12-31  green       0   
3       45.0      55.3      150    1.90      90 2024-03-10    red       1   
4       18.7      22.4      200    1.68      60 2024-07-22   blue       1   
5       60.2      45.1       10    1.72      68 2024-02-14  green       0   
6       33.8       NaN       50    1.85      85 2024-09-05    red       0   
7       41.3      38.7      100    1.78      74 2024-04-18   blue       1   
8       29.6      33.4      150    1.65      58 2024-11-30  green       0   
9       55.0      60.1      200    1.70      63 2024-08-12    red       1   
10      22.4      28.9       10    1.88      82 2024-05-25   blue       0   
11      38.1      42.6       50    1.80      78 2024-10-08  green       1   

**Polynomial features** are created by generating new features that are polynomial combinations of the existing features. This can be particularly useful when modeling non-linear relationships.

**Example:**

Suppose you have a single feature `x`. You can create polynomial features like `x^2`, `x^3`, etc., which may help in capturing the non-linear relationships in the data.

In [107]:
from sklearn.preprocessing import PolynomialFeatures
import pandas as pd

# Impute missing values in numerical_df before creating polynomial features
numerical_df_imputed = numerical_df.fillna(numerical_df.mean())

# Creating polynomial features (degree 2)
poly = PolynomialFeatures(degree=2, include_bias=False)
poly_features = poly.fit_transform(numerical_df_imputed)

# Convert the polynomial features back to a DataFrame
# The columns names will be based on the numerical_df columns
df_poly = pd.DataFrame(poly_features, columns=poly.get_feature_names_out(numerical_df_imputed.columns))

print(df_poly)

     feature1  feature2  feature  height  weight  target   age   income     x  \
0   25.500000   30.2000     10.0    1.60    55.0     1.0  25.0  30000.0   2.0   
1   32.100000   41.3425     50.0    1.75    80.0     0.0  35.0  50000.0   3.0   
2   37.374419   41.3425    100.0    1.82    72.0     0.0  45.0  70000.0   4.0   
3   45.000000   55.3000    150.0    1.90    90.0     1.0  55.0  90000.0   5.0   
4   18.700000   22.4000    200.0    1.68    60.0     1.0  28.0  35000.0   6.0   
5   60.200000   45.1000     10.0    1.72    68.0     0.0  32.0  48000.0   7.0   
6   33.800000   41.3425     50.0    1.85    85.0     0.0  40.0  62000.0   8.0   
7   41.300000   38.7000    100.0    1.78    74.0     1.0  38.0  55000.0   9.0   
8   29.600000   33.4000    150.0    1.65    58.0     0.0  30.0  41000.0  10.0   
9   55.000000   60.1000    200.0    1.70    63.0     1.0  42.0  68000.0   2.0   
10  22.400000   28.9000     10.0    1.88    82.0     0.0  27.0  32000.0   3.0   
11  38.100000   42.6000     

In this example, the original feature `x` is transformed into two features: `x` and `x^2`, allowing a model to capture quadratic relationships in the data.


---

### Target Encoding with Smoothing
**Target encoding** transforms categorical variables by replacing each category with the mean of the target variable for that category. While this method can capture the relationship between categorical features and the target, it may lead to overfitting, especially when some categories have few observations.

**Smoothing** addresses this issue by balancing the category-specific mean with the overall mean of the target variable. This approach ensures that categories with limited data are not assigned misleadingly extreme values.

#### Smoothing Formula
The smoothed target encoding for a category is calculated as:

$$
\text{Encoding} = \frac{n \cdot \text{Mean}_{\text{category}} + m \cdot \text{Mean}_{\text{global}}}{n + m}
$$

Where:

 - $ n $ is the number of observations for the category.
 - $ \text{Mean}_{\text{category}} $ is the mean target value for the category.
 - $ \text{Mean}_{\text{global}} $ is the overall mean target value across all categories.
 - $ m $ is the smoothing parameter that determines the weight of the global mean.

A higher $ m $ value increases the influence of the global mean, which is particularly useful for categories with few observations.

### Implementation in Python

Here's how you can implement smoothed target encoding using `pandas`

In [108]:
import pandas as pd

def target_encode_smooth(df, cat_col, target_col, m):
    # Compute the global mean
    global_mean = df[target_col].mean()

    # Compute the number of observations and the mean for each category
    agg = df.groupby(cat_col)[target_col].agg(['count', 'mean'])

    # Compute the smoothed means
    smooth = (agg['count'] * agg['mean'] + m * global_mean) / (agg['count'] + m)

    # Map the smoothed means to the original dataframe
    return df[cat_col].map(smooth)

This function calculates the smoothed target encoding for a specified categorical column. You can adjust the smoothing parameter $ m $ based on your dataset's characteristis.

---


More examples with **Feature engineering**.

**a. Creating New Features:**
- **Polynomial Features:** Creating polynomial combinations of existing features to capture non-linear relationships.

In [110]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2)
df_poly = poly.fit_transform(numerical_df_imputed)


- **Interaction Features:** Combining two or more features to capture interactions.

In [111]:
df['new_feature'] = df['feature1'] * df['feature2']


**b. Encoding Categorical Variables:**
- **One-Hot Encoding:** Converts categorical variables into a series of binary columns.

In [112]:
df_encoded = pd.get_dummies(df, columns=['categorical_column'])

- **Target Encoding:** Replaces categorical values with the mean of the target variable for that category.

In [113]:
df['encoded_column'] = df.groupby('categorical_column')['target'].transform('mean')


**c. Feature Selection:**
- **Correlation Analysis:** Identify features that are highly correlated with the target variable.

In [114]:
# Select only numerical columns for correlation
numerical_df = df.select_dtypes(include=[float, int])

# Calculate the correlation matrix
correlation_matrix = numerical_df.corr()

# Print the correlations with the 'target' variable, sorted in descending order
print(correlation_matrix['target'].sort_values(ascending=False))

target                    1.000000
encoded_column            0.842701
income                    0.348913
age_income_interaction    0.330753
age                       0.323837
height                    0.169898
feature2                  0.161201
weight                    0.144666
color_encoded             0.141165
feature                   0.135649
new_feature               0.132959
feature1                  0.123779
BMI                       0.094750
x                        -0.066391
month                    -0.169946
year                           NaN
Name: target, dtype: float64



Feature engineering is a powerful tool in the data preprocessing process. By creating new features, encoding categorical variables, generating interaction terms, and adding polynomial features, you can significantly enhance the performance of machine learning models. The effectiveness of feature engineering often depends on the domain knowledge and creativity of the practitioner, as it requires understanding the underlying data and the problem being solved.